server/service/src/pack_unit.rs

```diff
@@ -0,0 +1,176 @@
+use std::collections::HashMap;
+
+use repository::{
+    EqualFilter, PackUnitRow, PackUnitRowRepository, RepositoryError, StockLineFilter,
+    StockLineRepository, StockOnHandFilter, StockOnHandRepository, StorageConnection,
+};
+
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct PackUnit {
+    pub item_id: String,
+    pub most_used_pack_size_id: String,
+    pub pack_units: Vec<PackUnitRow>,
+}
+
+pub fn get_pack_units(
+    connection: &StorageConnection,
+    store_id: &str,
+) -> Result<Vec<PackUnit>, RepositoryError> {
+    let stock_lines = StockLineRepository::new(connection)
+        .query_by_filter(StockLineFilter::new(), Some(store_id.to_string()))?;
+    let stock_on_hand = StockOnHandRepository::new(connection).query(Some(
+        StockOnHandFilter::new().store_id(EqualFilter::equal_to(store_id)),
+    ))?;
+    let pack_units = PackUnitRowRepository::new(connection).load_all()?;
+
+    // Calculate the most used variant for each item and pack size by total number of packs
+    // if item has stock on hand else by empty line count. HashMap keys are (item_id, pack_size)
+    // for easier management of totals.
+    let mut total_number_of_packs: HashMap<(String, i32), f64> = HashMap::new();
+    let mut total_number_of_lines: HashMap<(String, i32), f64> = HashMap::new();
+    for stock_line in stock_lines {
```

Comment:
Thought about doing all this in an SQL view but might be too complicated? 🤔 idk

Reply:
we will need to test a large datafile, the login is already a slow process!
I'd prefer to have the logic in rust and keep the database clean - it's hard to maintain logic in two places in the codebase